okio/jvm/src/main/java/okio/Pipe.kt

```diff
@@ -66,25 +74,81 @@ class Pipe(internal val maxBufferSize: Long) {
           (buffer as Object).notifyAll() // Notify the source that it can resume reading.
         }
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { write(source, byteCount) } }
     }
 
     override fun flush() {
+      var delegate: Sink? = null
       synchronized(buffer) {
         check(!sinkClosed) { "closed" }
-        if (sourceClosed && buffer.size > 0L) throw IOException("source is closed")
+
+        foldedSink?.let {
+          delegate = it
+          return@synchronized
+        }
+
+        if (sourceClosed && buffer.size > 0L) {
+          throw IOException("source is closed")
+        }
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { flush() } }
     }
 
     override fun close() {
+      var delegate: Sink? = null
       synchronized(buffer) {
         if (sinkClosed) return
+
+        foldedSink?.let {
+          delegate = it
+          return@synchronized
+        }
+
         if (sourceClosed && buffer.size > 0L) throw IOException("source is closed")
         sinkClosed = true
         (buffer as Object).notifyAll() // Notify the source that no more bytes are coming.
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { close() } }
     }
 
     override fun timeout(): Timeout = timeout
+
+    private fun honorsTimeoutAndDeadline(delegateSink: Sink, block: Sink.() -> Unit) {
+      val sinkTimeoutNanos = this.timeout().timeoutNanos()
+      val delegateOriginalTimeoutNanos = delegateSink.timeout().timeoutNanos()
+      val minTimeoutNanos = when {
+        sinkTimeoutNanos == 0L -> delegateOriginalTimeoutNanos
+        delegateOriginalTimeoutNanos == 0L -> sinkTimeoutNanos
+        sinkTimeoutNanos < delegateOriginalTimeoutNanos -> sinkTimeoutNanos
+        else -> delegateOriginalTimeoutNanos
+      }
+
+      val sinkDeadline =
+        if (this.timeout().hasDeadline()) this.timeout().deadlineNanoTime() else null
```

Comment:
can you rewrite this without nulls ? Nulls mean boxes and boxes mean allocations

Reply:
actually even better – just define a method on `Timeout` that takes a lambda and does this internally. When you’re inside the timeout class you don’t have to do nearly as much awkward stuff to make the timeout/deadline stuff work